theurgy.lic

```diff
@@ -55,75 +130,357 @@ def buy_wine
   pause 1
   waitrt?
-  DRC.fix_standing
 end
 
-fput "get #{$settings.flint_lighter}", 'You get a', 'You are already'
-get_from_container('incense')
-while 'nothing happens' == DRC.bput('light my incense with flint', 'nothing happens', 'bursts into flames')
+reflectory = proc do
+  DRCT.walk_to '5988'
+  fput 'meditate'
   waitrt?
 end
+
+invoke_parchment = proc do
+  DRCT.walk_to('1230')
+  if 'but there is nothing in there like that' != DRC.bput("rummage /C parchment my #{$settings.theurgy_supply_container}", 'but there is nothing in there like that', 'pale golden prayer parchment')
+    get_from_container('golden parchment')
+    fput 'invoke my parchment'
+    waitfor 'You conclude'
+    put_in_container('golden parchment')
+  end
+end
+
+######################## CUSTOM TRAVEL ########################
+
+enter_altar = proc do
+  DRCT.walk_to('1164')
+  move 'go shack'
+  move 'go hole'
+  move 'west'
+  move 'up'
+  waitrt?
+  move 'go path'
+end
+
+exit_altar = proc do
+  move 'go path'
+  move 'go hole'
+  waitrt?
+  move 'east'
+  move 'go hole'
+  move 'out'
+end
+
+######################## MISC ########################
+
+def research_check
+  case find_current_project = DRC.bput("research status", / (\d+)% complete with a portion .*\./, 'not researching anything', 'project about')
+  when /not researching anything/
+    return
+  when /project about/
+    research_type = ["Fundamental", "Augmentation", "Stream", "Sorcery", "Utility", "Warding"]
+    $hold_research = research_type.find{ |$hold_research| find_current_project =~ /#{$hold_research}/i }
+    if $hold_research.nil?
+      return
+    end
+  when /complete with a portion/
+    echo find_current_project
+    research_type = ["Fundamental", "Augmentation", "Stream", "Sorcery", "Utility", "Warding"]
+    $hold_research = research_type.find{ |$hold_research| find_current_project =~ /#{$hold_research}/i }
+    if $hold_research.nil?
+      return
+    end
+    if find_current_project[1,2].to_i >= 60
+      wait_time = (101 - find_current_project[1,2].to_i) * 3
+      Flags.reset('research_done')
+      pause (wait_time)
+    else
+      echo "stopping research"
+    end
+    fput 'prep bless'
+  end
+  echo $hold_research
+  return $hold_research
+end
+
+def controller(method)
+  echo Flags['theurgy-commune']
+  echo Flags['theurgy-eluned']
+  echo Flags['theurgy-tasmine']
+  echo Flags['theurgy-truffenyi']
+  echo Flags['theurgy-kertigen']
+  if Flags['theurgy-commune']
+    Flags.reset('theurgy-commune')
+    next_commune
+  end
+  method.call
+end
+
+def next_commune
+  if Flags['theurgy-eluned'] == false && DRStats.circle > 3 && DRSkill.getrank('Outdoorsmanship') >= 20 && DRSkill.getxp('Theurgy') <=31
```

Comment:
Unless I'm wrong we won't be running any of these communes twice, we could just determine the valid list of communes at the beginning of the file by pushing each one into the array if the requirements are met, then we can just call `@commune_list.shift.call if @commune_list.any?` in the next commune method. I think we're kind of setting ourselves up for failure by trying to make this training more modular than it needs to be. For example commune eluned has the implicit assumption that you are somewhere that you can forage dirt, but it just relies on that happening through loose structure instead of just doing commune_eluned once at the beginning when you know you're in a dirt room.